File: it-agents/skills/code-architect/references/code-examples/cpp/Command.cpp
// Command — encapsulate a request as an object, support undo.
#include <iostream>
#include <memory>
#include <stack>
#include <string>
#include <vector>

// Receiver holds the state operated on.
class Document {
    std::string text;
public:
    void append(const std::string& s) { text += s; }
    void erase(size_t n) { text.erase(text.size() - n); }
    const std::string& content() const { return text; }
};

// Command interface.
class Command {
public:
    virtual ~Command() = default;
    virtual void execute() = 0;
    virtual void undo() = 0;
};

class AppendCommand : public Command {
    Document& doc;
    std::string fragment;
public:
    AppendCommand(Document& d, std::string s) : doc(d), fragment(std::move(s)) {}
    void execute() override { doc.append(fragment); }
    void undo() override { doc.erase(fragment.size()); }
};

// Invoker keeps a history for undo.
class Invoker {
    std::stack<std::unique_ptr<Command>> history;
public:
    void run(std::unique_ptr<Command> cmd) {
        cmd->execute();
        history.push(std::move(cmd));
    }
    void undo() {
        if (history.empty()) return;
        history.top()->undo();
        history.pop();
    }
};

int main() {
    Document doc;
    Invoker invoker;
    invoker.run(std::make_unique<AppendCommand>(doc, "Hello"));
    invoker.run(std::make_unique<AppendCommand>(doc, ", World"));
    std::cout << doc.content() << "\n";
    invoker.undo();
    std::cout << doc.content() << "\n";
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/rust/command.rs
// Command — encapsulate a request as an object, decoupling sender from receiver
// and enabling features like undo via an inverse operation.

// Receiver: the object commands act upon.
struct Document {
    content: String,
}

trait Command {
    fn execute(&self, doc: &mut Document);
    fn undo(&self, doc: &mut Document);
}

struct Append {
    text: String,
}
struct Clear {
    saved: std::cell::RefCell<String>,
}

impl Command for Append {
    fn execute(&self, doc: &mut Document) {
        doc.content.push_str(&self.text);
    }
    fn undo(&self, doc: &mut Document) {
        let new_len = doc.content.len() - self.text.len();
        doc.content.truncate(new_len);
    }
}

impl Command for Clear {
    fn execute(&self, doc: &mut Document) {
        *self.saved.borrow_mut() = std::mem::take(&mut doc.content);
    }
    fn undo(&self, doc: &mut Document) {
        doc.content = self.saved.borrow().clone();
    }
}

fn main() {
    let mut doc = Document { content: String::new() };
    let mut history: Vec<Box<dyn Command>> = Vec::new();

    for cmd in [
        Box::new(Append { text: "Hello".into() }) as Box<dyn Command>,
        Box::new(Append { text: ", World".into() }),
        Box::new(Clear { saved: Default::default() }),
    ] {
        cmd.execute(&mut doc);
        println!("after exec: '{}'", doc.content);
        history.push(cmd);
    }

    while let Some(cmd) = history.pop() {
        cmd.undo(&mut doc);
        println!("after undo: '{}'", doc.content);
    }
}
